src/validation.cpp

```diff
@@ -4698,18 +4698,31 @@ void ChainstateManager::CheckBlockIndex()
         return;
     }
 
-    // Build forward-pointing map of the entire block tree.
+    // Build forward-pointing data structure for the entire block tree.
+    // For performance reasons, indexes of the best header chain are stored in a vector.
+    // All remaining blocks are stored in a multimap.
+    CBlockIndex* index_walk{m_best_header};
+    const int best_hdr_height{m_best_header->nHeight};
+    std::vector<CBlockIndex*> best_hdr_chain(best_hdr_height + 1);
```

Comment:
I didn't do that so that the speedup is independent from the chain, and also applies to situations where we have all the headers, but `m_chain` is behind (IBD, reindex). Which seemed important because these are the exact situations where `CheckBlockIndex` will be called a lot and slows `bitcoind` down most noticeably.